source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -551,20 +556,27 @@ void Filter::onGenericPoolReady(StreamInfo::StreamInfo* info,
                                 Upstream::HostDescriptionConstSharedPtr& host,
                                 const Network::ConnectionInfoProvider& address_provider,
                                 Ssl::ConnectionInfoConstSharedPtr ssl_info) {
+  StreamInfo::UpstreamInfo& upstream_info = *getStreamInfo().upstreamInfo();
   upstream_ = std::move(upstream);
   generic_conn_pool_.reset();
   read_callbacks_->upstreamHost(host);
-  StreamInfo::UpstreamInfo& upstream_info = *getStreamInfo().upstreamInfo();
+  // No need to set information using address_provider in case routing via Router::UpstreamRequest
+  // because in that case, information is already set by the
+  // Router::UpstreamRequest::onPoolReady() method before reaching here.
+  if ((!Runtime::runtimeFeatureEnabled(
```

Comment:
if we're assuming things are set or not set is it going to cause problems if the flag flips mid session?  I'd think we'd need to latch this value and  use it consistently for a given connection